libp2p/network/health/monitor.py

```diff
@@ -0,0 +1,365 @@
+        """
+        try:
+            # If there are active streams, avoid intrusive ping; assume healthy
+            if len(conn.get_streams()) > 0:
+                return True
+
+            # Use a timeout for the ping
+            with trio.move_on_after(self.config.ping_timeout):
+                # Create a throwaway stream and immediately reset it to avoid
+                # affecting muxer stream accounting in tests
+                stream = await conn.new_stream()
+                try:
+                    await stream.reset()
+                finally:
+                    # Best-effort close in case reset was a no-op
+                    try:
+                        await stream.close()
+                    except Exception:
+                        pass
+                return True
+
+        except Exception as e:
+            logger.debug(f"Ping failed for connection: {e}")
+
+        return False
+
+    def _should_replace_connection(self, peer_id: ID, conn: INetConn) -> bool:
+        """Determine if a connection should be replaced based on health metrics."""
+        if not self._has_health_data(peer_id, conn):
+            return False
+
+        health = self.swarm.health_data[peer_id][conn]
+        config = self.config
+
+        # Check various health thresholds
+        unhealthy_reasons = []
+
+        if health.health_score < config.min_health_threshold:
+            unhealthy_reasons.append(f"low_health_score={health.health_score:.2f}")
+
+        if health.ping_latency > config.max_ping_latency:
+            unhealthy_reasons.append(f"high_latency={health.ping_latency:.1f}ms")
+
+        if health.ping_success_rate < config.min_ping_success_rate:
+            unhealthy_reasons.append(f"low_success_rate={health.ping_success_rate:.2f}")
+
+        if health.failed_streams > config.max_failed_streams:
+            unhealthy_reasons.append(f"too_many_failed_streams={health.failed_streams}")
+
+        if unhealthy_reasons:
+            # If connection is in active use (streams open), do not replace
+            try:
+                if len(conn.get_streams()) > 0:
+                    return False
+            except Exception:
+                pass
+
+            # Require N consecutive unhealthy evaluations before replacement
+            health.consecutive_unhealthy += 1
+            if health.consecutive_unhealthy >= getattr(
+                config, "unhealthy_grace_period", 1
+            ):
+                logger.info(
+                    f"Connection to {peer_id} marked for replacement: "
+                    f"{', '.join(unhealthy_reasons)}"
+                )
+                health.consecutive_unhealthy = 0
+                return True
+            return False
+        else:
+            # Reset counter when healthy again
+            if hasattr(health, "consecutive_unhealthy"):
+                health.consecutive_unhealthy = 0
+
+        return False
+
+    async def _replace_unhealthy_connection(
```

Comment:
In this method, we first check if removing the bad connection would drop us below `min_connections_per_peer`, and only replace that bad connection if it won't.

But what if we are sitting at or below that threshold? Then we would just sit with a bad connection and never try to replace it. Would it make sense to try to try to establish a new connection with the peer first, and then drop the bad connection if it's successful?